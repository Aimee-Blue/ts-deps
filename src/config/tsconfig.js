// @ts-check
const { options } = require('../options');
const { trimPathSeparator } = require('../helpers');
const defaults = require('../defaults');

/**
 * @param {Partial<ITsConfigParams>} paramsRaw
 */
const tsConfig = (paramsRaw = {}) => {
  const opts = options();
  const { aliases, baseConfigLocation } = {
    aliases: opts.aliases,
    baseConfigLocation:
      opts.baseTsConfigLocation ||
      './node_modules/ts-deps/lib/config/tsconfig.default.json',
    ...paramsRaw,
  };
  const src = defaults.rootDir;
  const lib = defaults.outDir;
  const paths =
    aliases &&
    Object.keys(aliases).reduce((acc, key) => {
      const alias = trimPathSeparator(aliases[key]);
      return {
        ...acc,
        [key]: [alias],
        [`${key}/*`]: [`${alias}/*`],
      };
    }, {});

  return {
    extends: baseConfigLocation,
    compilerOptions: {
      baseUrl: '.',
      paths,
      outDir: lib,
    },
    include: [src],
  };
};

/**
 * @param {Partial<ITsConfigParams>} paramsRaw
 */
const tsConfigDeclarations = (paramsRaw = {}) => {
  /**
   * @type {ReturnType<typeof tsConfig> & { exclude?: string[] }}
   */
  const defConfig = tsConfig(paramsRaw);

  return {
    ...defConfig,
    compilerOptions: {
      ...defConfig.compilerOptions,
      allowJs: false,
      noEmit: false,
      declaration: true,
      emitDeclarationOnly: true,
    },
    exclude: [
      ...((Array.isArray(defConfig.exclude) && defConfig.exclude) || []),
      defaults.unitTestsGlob,
      defaults.integrationTestsGlob,
    ],
  };
};

module.exports = {
  tsConfig,
  tsConfigDeclarations,
};
