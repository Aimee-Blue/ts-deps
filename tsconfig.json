{
  "extends": "./src/config/tsconfig.default.json",
  "compilerOptions": {
    "paths": { "ts-deps": ["./src"], "ts-deps/*": ["./src/*"] },
    "outDir": "lib"
  },
  "include": ["src"]
}
