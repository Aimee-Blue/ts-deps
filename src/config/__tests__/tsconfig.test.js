import { tsConfig, tsConfigDeclarations } from '../tsconfig';

describe('given no options', () => {
  it('should work', () => {
    const defaultConfig = tsConfig();
    expect(defaultConfig).toMatchSnapshot();
  });
});

describe('given options', () => {
  it('should work', () => {
    const defaultConfig = tsConfig({
      aliases: {
        '@shared': '../../long/path/src',
      },
      baseConfigLocation: 'node_modules/my-custom-lib/tsconfig.json',
    });
    expect(defaultConfig).toMatchSnapshot();
  });
});

describe('given declarations', () => {
  it('should work', () => {
    const defaultConfig = tsConfigDeclarations({
      aliases: {
        '@shared': '../../long/path/src',
      },
      baseConfigLocation: 'node_modules/my-custom-lib/tsconfig.json',
    });
    expect(defaultConfig).toMatchSnapshot();
  });
});
