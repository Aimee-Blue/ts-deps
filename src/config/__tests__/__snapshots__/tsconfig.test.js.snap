// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given declarations should work 1`] = `
Object {
  "compilerOptions": Object {
    "allowJs": false,
    "baseUrl": ".",
    "declaration": true,
    "emitDeclarationOnly": true,
    "jsx": "react",
    "noEmit": false,
    "outDir": "lib",
    "paths": Object {
      "@shared": Array [
        "../../long/path/src",
      ],
      "@shared/*": Array [
        "../../long/path/src/*",
      ],
    },
  },
  "exclude": Array [
    "src/**/__tests__/**",
    "src/__integration-tests__/**",
  ],
  "extends": "node_modules/my-custom-lib/tsconfig.json",
  "include": Array [
    "src",
  ],
}
`;

exports[`given no options should work 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "jsx": "react",
    "outDir": "lib",
    "paths": Object {
      "ts-deps": Array [
        "./src",
      ],
      "ts-deps/*": Array [
        "./src/*",
      ],
    },
  },
  "extends": "./src/config/tsconfig.default.json",
  "include": Array [
    "src",
  ],
}
`;

exports[`given options should work 1`] = `
Object {
  "compilerOptions": Object {
    "baseUrl": ".",
    "jsx": "react",
    "outDir": "lib",
    "paths": Object {
      "@shared": Array [
        "../../long/path/src",
      ],
      "@shared/*": Array [
        "../../long/path/src/*",
      ],
    },
  },
  "extends": "node_modules/my-custom-lib/tsconfig.json",
  "include": Array [
    "src",
  ],
}
`;
